cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -948,11 +949,17 @@ func (csr *ClusterStateRegistry) getCloudProviderNodeInstances() (map[string][]c
 }
 
 // Calculates which of the existing cloud provider nodes are not registered in Kubernetes.
-func getNotRegisteredNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances map[string][]cloudprovider.Instance, time time.Time) []UnregisteredNode {
+func (csr *ClusterStateRegistry) getNotRegisteredNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances map[string][]cloudprovider.Instance, time time.Time) []UnregisteredNode {
 	registered := sets.NewString()
 	for _, node := range allNodes {
-		registered.Insert(node.Spec.ProviderID)
+		if csr.cloudProvider.Name() == cloudprovider.JujuProviderName {
+			hostname := node.Labels[juju.HostnameLabel]
+			registered.Insert(hostname)
+		} else {
+			registered.Insert(node.Spec.ProviderID)
+		}
```

Comment:
i *know* this works technically.  But i almost would rather do the "right thing" and update the Interface object to have a new required field that does this correctly.  Even if it means EVERY provider has to update.  This is such a sad state.  Maybe create an issue on the upstream product so we can track why this atrocity is here.

Reply:
let me see what I can do, I can update the interface, and with conditional compilation we might be able to get away with not updating every single cloud provider from here to eternity.

Reply:
Ive updated the PR to instead fix the provider ID mismatch by querying k8s for the provider IDs via the hostname provided by the juju side of things. The provider ID is tracked with each juju unit along with the hostname, and as part of the refresh method it is kept up to date if things change. 

When [this bug](https://bugs.launchpad.net/juju/+bug/1968092) is resolved all of this will be much easier as we wont have to micromanage every single unit in order to determine its lifecycle (currently there is no way to tell via juju status if a non active/non idle unit is coming up or being destroyed, so we have to meticulously  keep track of each little change in order to determine the actual state of the lifecycle a unit is in.